ambry-network/src/main/java/com.github.ambry.network/http2/Http2NetworkClient.java

```diff
@@ -17,38 +17,83 @@
 import com.github.ambry.commons.SSLFactory;
 import com.github.ambry.config.Http2ClientConfig;
 import com.github.ambry.network.NetworkClient;
+import com.github.ambry.network.NetworkClientErrorCode;
 import com.github.ambry.network.RequestInfo;
 import com.github.ambry.network.ResponseInfo;
+import com.github.ambry.utils.SystemTime;
+import io.netty.channel.Channel;
 import io.netty.channel.EventLoopGroup;
 import io.netty.channel.nio.NioEventLoopGroup;
 import io.netty.channel.pool.ChannelPool;
 import io.netty.channel.pool.ChannelPoolMap;
+import io.netty.util.AttributeKey;
 import java.net.InetSocketAddress;
 import java.util.List;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import java.util.ArrayList;
+import java.util.Queue;
 import java.util.Set;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 
 /**
  * A HTTP/2 implementation of {@link NetworkClient}.
- * TODO: More implementation details will be added in next PR.
  */
 public class Http2NetworkClient implements NetworkClient {
   private static final Logger logger = LoggerFactory.getLogger(Http2NetworkClient.class);
   private final EventLoopGroup eventLoopGroup;
   private final ChannelPoolMap<InetSocketAddress, ChannelPool> pools;
+  private Http2ClientResponseHandler http2ClientResponseHandler;
+  public static AttributeKey<RequestInfo> REQUEST_INFO = AttributeKey.newInstance("RequestInfo");
 
   public Http2NetworkClient(Http2ClientMetrics http2ClientMetrics, Http2ClientConfig http2ClientConfig,
       SSLFactory sslFactory) {
     this.eventLoopGroup = new NioEventLoopGroup(http2ClientConfig.http2NettyEventLoopGroupThreads);
     this.pools = new Http2ChannelPoolMap(sslFactory, eventLoopGroup, http2ClientConfig);
+    this.http2ClientResponseHandler = new Http2ClientResponseHandler();
   }
 
   @Override
   public List<ResponseInfo> sendAndPoll(List<RequestInfo> requestsToSend, Set<Integer> requestsToDrop,
       int pollTimeoutMs) {
-    return null;
+    long startTime = SystemTime.getInstance().milliseconds();
+    List<ResponseInfo> readyResponseInfo = new ArrayList<>();
+    // Send request
+    for (RequestInfo requestInfo : requestsToSend) {
+      requestInfo.streamAcquireTime = SystemTime.getInstance().milliseconds();
+      this.pools.get(InetSocketAddress.createUnresolved(requestInfo.getHost(), requestInfo.getPort().getPort()))
+          .acquire()
```

Comment:
can we add pollTimeout at acquire method?

Reply:
The pollTimeout is used for response waiting and timeout.